cmd/tail.go

```diff
@@ -0,0 +1,154 @@
+package cmd
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/Benchkram/errz"
+	"github.com/hpcloud/tail"
+	"github.com/puppetlabs/wash/api/client"
+	apitypes "github.com/puppetlabs/wash/api/types"
+	"github.com/puppetlabs/wash/config"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+)
+
+func tailCommand() *cobra.Command {
+	tailCmd := &cobra.Command{
+		Use:   "tail -f <file> [<file>...]",
+		Short: "Displays new output of files or resources that support the stream action",
+	}
+
+	tailCmd.Flags().BoolP("follow", "f", false, "Follow new output (required)")
+	if err := viper.BindPFlag("follow", tailCmd.Flags().Lookup("follow")); err != nil {
+		log.Print(err)
+	}
+
+	tailCmd.RunE = toRunE(tailMain)
+
+	return tailCmd
+}
+
+type line struct {
+	tail.Line
+	source string
+}
+
+type lineWriter struct {
+	name string
+	out  chan line
+}
+
+func (w lineWriter) Write(b []byte) (int, error) {
+	s := strings.TrimSuffix(string(b), "\n")
+	w.out <- line{Line: tail.Line{Text: s, Time: time.Now()}, source: w.name}
+	return len(b), nil
+}
+
+func tailMain(cmd *cobra.Command, args []string) exitCode {
+	follow := viper.GetBool("follow")
+	if !follow {
+		log.Print("Please use -f, other operations are not yet supported")
+		return exitCode{1}
+	}
+
+	// If no paths are declared, try to stream the current directory/resource
+	if len(args) == 0 {
+		args = []string{"."}
+	}
+
+	conn := client.ForUNIXSocket(config.Socket)
+	agg := make(chan line)
+
+	// Separate paths into streamable resources and files
+	var files []string
+	for _, path := range args {
+		apiPath, err := client.APIKeyFromPath(path)
+		if err != nil {
+			// Not a resource
+			files = append(files, path)
+			continue
+		}
+
+		if stream, err := conn.Stream(apiPath); err == nil {
+			defer func() { errz.Log(stream.Close()) }()
+			// Start copying the stream to the aggregate channel
+			go func(src string) {
+				_, err := io.Copy(lineWriter{name: src, out: agg}, stream)
+				if err != nil {
+					agg <- line{Line: tail.Line{Time: time.Now(), Err: err}, source: src}
+				}
+			}(path)
+		} else {
+			if errObj, ok := err.(*apitypes.ErrorObj); ok {
+				if errObj.Kind == apitypes.UnsupportedAction {
+					// The resource exists but does not support the streaming action, try to read it as a file.
+					files = append(files, path)
+				} else {
+					log.Print(errObj.Msg)
+				}
+			} else {
+				log.Print(err)
+			}
+		}
+	}
+
+	// Set Location so we start streaming at the end of the file
+	tailConf := tail.Config{
+		Follow:   true,
+		Location: &tail.SeekInfo{Offset: 0, Whence: 2},
+		Logger:   tail.DiscardingLogger,
+	}
+	for _, path := range files {
+		if finfo, err := os.Stat(path); err != nil {
+			log.Print(err)
+			continue
+		} else if finfo.IsDir() {
```

Comment:
A comment explaining why we're doing this check would be useful, b/c `tail` working on a directory is unintuitive.

Reply:
Oh, I was talking about commenting that the `tail` package works on directories -- that was the unintuitive part and the reason for the `IsDir()` check.

Reply:
The comment at line 107 doesn't clarify that?

Reply:
Oh I see. I'll specifically mention the `tail` package behavior.

Reply:
It might actually be waiting for the file to exist. I wonder if I should still call `TailFile` on these.

Reply:
Up to you. What you have here seems OK to me.